### ISSUE TITLE
second phase of indexing should be better parallelized

### ISSUE THREAD

**Issue by vladak** [oracle/opengrok#1451]
When performing full index of multiple projects, where one of the project's cardinality is big (i.e. the number files in it is several magnitudes bigger than the others), it often happens that initially the machine performing the indexing is fully loaded (thanks to indexing of projects happens in parallel) and in the end there is just one project that is being indexed. At that point there is basically just the thread processing the output from `ctags` running. This makes overall index slow.
I think there used to be a Bugzilla bug where that describes this behavior however I could not find it.

The behavior stems from the fact that each project indexer runs its own ctags process. The stack looks like this:

```
IndexDatabase.update()
IndexDatabase.indexDown()
IndexDatabase.addFile()
AnalyzerGuru.populateDocument()
AbstractSourceCodeAnalyzer.analyze()
PlainAnalyzer.analyze()
Ctags.doCtags()
```

There could be a pool of ctags processes with dispatcher (which would be aware of which ctags process is busy/free). This probably means the parallelization done in `doIndexerExecution()` (submits new job with `IndexDatabase.update()` for each project) would have to reworked to make parallelization happening on file level - basically each `addFile()` would submit a request that would be processed by a thread and one of the ctags processes. The projects would be processed sequentially by single thread however within each projects files would be processed in parallel. Basically the project level parallelism would be replaced by file-level parallelism.

This might also require creating the directory structure for xrefs before the parallel job starts running - depends on the order in which the files would be added to queue.

**Comment by tulinkry** [comment]
The #460 bug describes the same effect for the history part.

**Comment by vladak** [comment]
The ideas were implemented by #1936, closing.
